### ISSUE TITLE
Runbook for executing Linky link script.

### ISSUE THREAD

**Issue by ndouglas** [department-of-veterans-affairs/va.gov-cms#9937]
## Description
In #9936 , we finalized the script for creating Linky links on production.

However, this script will need to run for a long time and should probably be run in multiple segments rather than in a single command.

It should be run outside of normal editor work hours, just to minimize confusion, content locks, etc.

Each command run should be run on staging prior to production and it should be confirmed to execute completely without error.

## Acceptance Criteria
- [ ] A runbook has been created that details the steps to execute on BRD environments.

**Comment by EWashb** [comment]
@ndouglas do you have any idea of how long it should run? I know you said a long time. So maybe a month? More?

**Comment by ndouglas** [comment]
After rewriting/reorganizing the script and only touching particular nodes for a certain content type, the runtime is less of an issue.

The current time this takes on Tugboat is ~45 minutes.  I expect this might take a little longer on production -- possibly 60-75 minutes -- because there is some added latency for every database operation because the database server is a separate machine, somewhere else in AWS.

And of course I want to run this on staging first, then on production.

So this would probably look something like this:

- [ ] Day 1: run script on staging for all content types, see if anything bad happens.  See what the runtime is.
- [ ] Day 2: if runtime isn't that long, run script on prod for all content types -or- if it's long, run script on prod for 1/2 of the content, divided by length.
- [ ] Day 3 (optional): Run script on prod for remaining half of content.

I think what I'm going to do, though, is figure out the actual node updates for some of the other content types that have low cost-benefit ratios (i.e. not many linkies created, but it took a long time to do that), and then only save those particular content types.

That should cut the total runtime of the script substantially (at the risk of _possibly_ only having 99% of upfront linkies created, not 100%).

The runbook would probably look the same, though.

So **TL;DR**: I'm thinking three evenings, no more than that.  Probably two.

**Comment by EWashb** [comment]
Ah thank you for the detailed update. I was thinking running it several times over the course of a month would give us good data (why I said a month), but I see that my previous comment was confusing!

That all makes perfect sense and looking to see which issue to bring into sprint 66. Do you think those individual nodes need their own discovery ticket or this one is fine?

**Comment by ndouglas** [comment]
I already had a setting in the script, so I just switched it on and the script will say, e.g.:

```
Processing content type vamc_operating_status_and_alerts...
Found 139 nodes...
Loaded 50 nodes as chunk 0
Created 3 new linkies updating node 2370...
Loaded 50 nodes as chunk 1
Created 12 new linkies updating node 16281...
Loaded 39 nodes as chunk 2
Finished content type ... 15 new linky entities added in 8 seconds (1.875 links/second)
```

So there's 139 nodes with this content type.  Every single node is loaded and saved, which creates linkies (if appropriate). 

With this switch turned on, I count the existing linkies before and after saving each node.  If the number of linkies is increased, then it mentions how many linkies were created by saving the node.  And we see that 15 were created in 8 seconds.  Probably not worth optimizing further.

Some content types, though, for instance `health_care_region_detail_page`, it's just not worth it to go through and save every node because there just weren't very many qualifying links that would get turned into linkies.  There were 5,256 nodes, but only 47 linkies.  It ended up taking **three hours**, as the script was originally coded, to save those 47 linkies!

So I went through and copied the actual node IDs for `health_care_region_detail_page` and changed the script to _only_ save the nodes that actually created linkies.  Now it takes **9 seconds** to do the same amount of work.

So I'm just running the same script again, with that switch flipped on, and I'll see if there's another content type or two worth doing the same thing with.  If so, I'll copypasta their node IDs too.

**TL;DR**: Nah, not worth another discovery ticket.

**Comment by EWashb** [comment]
This is great, Nate!  Nice work and thank you for the amazing explanation. I'll officially pull this into next sprint unless you feel like it's getting done in 65. Totally up to you.

**Comment by EWashb** [comment]
@ndouglas did you get the issues you were discussing in scrum today worked out?

**Comment by ndouglas** [comment]
@EWashb Yeah, everything's groovy.  It's just a really big PR and aging 😩

**Comment by EWashb** [comment]
@ndouglas sad times. I just wanted to make sure you weren't blocked anymore

**Comment by ndouglas** [comment]
Staging keeps getting reset, so I'll need to do this on CMS-Test.

**Comment by ndouglas** [comment]
```
PHP Fatal error:  Allowed memory size of 4294967296 bytes exhausted (tried to allocate 1310720 bytes) in /var/www/cms/docroot/core/lib/Drupal/Core/Cache/CacheTagsChecksumTrait.php on line 131

real    31m46.098s
user    12m57.278s
sys     1m2.905s
```

I think I'll need to divide `regional_health_care_service_des` into chunks.

**Comment by ndouglas** [comment]
To be safe, I'll run it on the following content types first:

```
  campaign_landing_page
  page
  vet_center
  office
  news_story
  press_release
```

They're comparatively small.

**Comment by ndouglas** [comment]
Then I'll do:

```
  full_width_banner_alert
  health_care_local_facility
  event
  q_a
  press_releases_listing
```

They're comparatively small.

**Comment by ndouglas** [comment]
Doing the rest minus `regional_health_care_service_des`:

```
  health_care_region_detail_page
  vet_center_facility_health_servi
  vamc_system_policies_page
  vamc_operating_status_and_alerts
  person_profile
```

**Comment by ndouglas** [comment]
I split the `regional_health_care_service_des` eight ways to avoid the memory issue I ran into on staging.  It succeeded.

```
real    39m45.265s
user    12m41.067s
sys     0m48.048s
```

There were [some broken link errors](https://dsva.slack.com/archives/C030F5WV2TF/p1665629151182319) involving Linky in a concurrent content build, but I think they're a result of a race condition.  I'll see in the morning, I guess, on the next content build.

**Comment by ndouglas** [comment]
Well, I _guess_ this can be closed 😑
